### PR TITLE
Update DPoS library to validate seedReveal value - Closes #4937

### LINKED ISSUE #4937: Update DPoS library to validate seedReveal value
### Description
- Update DPoS library to verify seedReveal value following https://github.com/LiskHQ/lips/blob/master/proposals/lip-0022.md#validating-new-block-header-property.
- `isDPoSProtocolComplient(blockHeader: BlockHeader, store: StateStore) => boolean`.
If it’s invalid seed reveal, the reward should be changed to 0 in the block_processor.
- Update `BlockHeader` to include `seedReveal`
- It should use the lastBlockHeaders in the stateStore.consensus.lastBlockHeaders

### Motivation
In order to encourage inserting correct seed reveal, the value has to be checked
### Acceptance Criteria
- Test case where seed reveal does not match with the previous one
- Test case where seed reveal does match with previous one
- Test case where delegate did not forge for last 2 rounds

### Related issues
#4915

### Additional Information
- `isBFTProtocolComplient` already exists and it should follow similar usage

### PR DESCRIPTION
**Pull request by ManuGowda**
### What was the problem?

This PR resolves #4937 

### How was it solved?
- Implemented `isDPoSProtocolCompliant` function in `dpos` library to check for validity of the block
- Followed the specification and implemented https://github.com/LiskHQ/lips/blob/master/proposals/lip-0022.md#validating-new-block-header-property
- Block processor to check for `isDPoSProtocolCompliant` during block creation and validation
### How was it tested?

- Implemented unit test

### CODE REVIEW

File: elements/lisk-dpos/src/dpos.ts [hunk 1]
```diff
@@ -208,4 +208,34 @@ export class Dpos {
...
+		}
+
+		const { seedReveal: previousBlockSeedReveal } = delegateForgedBlocks.slice(
+			0,
+			1,
+		)[0];
+		const { seedReveal: newBlockSeedReveal } = blockHeader;
+
+		// Check if last block seedReveal is not a preimage of new block
+		if (previousBlockSeedReveal !== newBlockSeedReveal) {
```
Comment (shuse2) [inline elements/lisk-dpos/src/dpos.ts:234]:
```suggestion
		if (!hexToBuffer(previousBlockSeedReveal).equals(hash(hexToBuffer(newBlockSeedReveal)).slice(0, 16)) {
```
Where 0, and 16 is defined in #4939

File: elements/lisk-dpos/src/dpos.ts [hunk 2]
```diff
@@ -208,4 +208,34 @@ export class Dpos {
 			delegateListRoundOffset,
 		});
 	}
+
+	// tslint:disable-next-line: prefer-function-over-method
+	public async isDPoSProtocolCompliant(
```
Comment (shuse2) [inline elements/lisk-dpos/src/dpos.ts:213]: also, can you add this usage of this function in the block_procoessor_v2 where BFTProtocolCompliant is called?

File: elements/lisk-dpos/src/dpos.ts [hunk 3]
```diff
@@ -208,4 +208,34 @@ export class Dpos {
...
+		const delegateForgedBlocks = lastBlockHeaders.filter(
+			block => block.generatorPublicKey === blockHeader.generatorPublicKey,
+		);
+
+		if (!delegateForgedBlocks.length) {
+			// If the forger din't forge any block in the last three rounds
+			return true;
+		}
+
+		const { seedReveal: previousBlockSeedReveal } = delegateForgedBlocks.slice(
```
Comment (shuse2) [inline elements/lisk-dpos/src/dpos.ts:227]: i think you don't need to slice here?

File: elements/lisk-dpos/src/dpos.ts (hunk 1 above)
Comment (shuse2) [reply elements/lisk-dpos/src/dpos.ts:234]: Also, please update the relevant test

File: elements/lisk-dpos/src/dpos.ts (hunk 2 above)
Comment (ManuGowda) [reply elements/lisk-dpos/src/dpos.ts:213]: u mean a comment?

Comment (shuse2) [reply elements/lisk-dpos/src/dpos.ts:213]: I mean in this PR, block_processor_v2 should use this to check the dpos compliance and update/validate the reward

File: elements/lisk-dpos/src/dpos.ts [hunk 4]
```diff
@@ -208,4 +208,34 @@ export class Dpos {
...
+			0,
+			1,
+		)[0];
+		const { seedReveal: newBlockSeedReveal } = blockHeader;
+
+		// Check if last block seedReveal is not a preimage of new block
```
Comment (ishantiw) [inline elements/lisk-dpos/src/dpos.ts:233]: I think it should be opposite if I understood it correctly,
`// Check if the new block seed reveal is not a preimage of the last block seed reveal`

Comment (ishantiw) [reply elements/lisk-dpos/src/dpos.ts:233]: also, I think it sounds like a correct scenario if it's not equal, maybe emphasis on what it should be,
`New block seed reveal should be a preimage of the last block seed reveal`

File: elements/lisk-dpos/src/dpos.ts [hunk 5]
```diff
@@ -208,4 +209,34 @@ export class Dpos {
...
 		});
 	}
+
+	// This function is used in block_processor_v2 to check the dpos compliance and update/validate the reward
+	// tslint:disable-next-line: prefer-function-over-method
+	public async isDPoSProtocolCompliant(
+		blockHeader: BlockHeader,
+		store: StateStore,
+	): Promise<boolean> {
+		const delegateForgedBlocks = store.consensus.lastBlockHeaders.filter(
```
Comment (shuse2) [inline elements/lisk-dpos/src/dpos.ts:219]: this line needs to reflect
>  in the previous round or in the same round of B

from https://github.com/LiskHQ/lips/blob/master/proposals/lip-0022.md#validating-new-block-header-property

**Pull request merged**